### ISSUE TITLE
layer: Softmax backward (FLOAT32/SYM) never checks loss/propLoss element counts; SYM arm declares zero-length VLAs

### ISSUE THREAD

**Issue by LeoBuron** [es-ude/OnDeviceTraining#457]
## Problem

`softmaxBackwardSymInt32` converts all three wires into float stack scratch. It sizes all three buffers from the **input's** element count, and it declares them before any geometry or count check:

- **(a) Zero-element input:** it declares three zero-length VLAs. That is UB under C11 6.7.6.2p5, and clang's UBSan reports it as `vla-bound`. This was deferred during #152 as ruling R4 (spec §8).
- **(b) No check on loss/propLoss counts:** nothing compares their element counts with the input's. A longer loss or propLoss overflows the stack scratch. A shorter one produces silently wrong or truncated dx.

`softmaxBackwardFloat` is missing the same count gate. It works directly in the caller's buffers (heap for `initTensor`-built wires), so there a shorter loss is read out of bounds and a shorter propLoss is written out of bounds.

## Evidence (develop 3f4d7c31)

`src/layer/Softmax.c:345-378`:

```c
size_t inputSize = calcNumberOfElementsByTensor(input);                      // :346
uint8_t inputFloatData[inputSize * sizeof(float)];                          // :351
uint8_t lossFloatData[inputSize * sizeof(float)];                           // :358
setTensorValuesForConversion(lossFloatData, &lossFloatQ, loss, &lossFloat);
convertTensor(loss, &lossFloat);                                            // :360
uint8_t propLossFloatData[inputSize * sizeof(float)];                       // :365
convertTensor(propLoss, &propLossFloat);                                    // :367
softmaxRowGeometry(input, "Softmax backward", &rows, &rowLen);              // :373  first geometry check
convertTensor(&propLossFloat, propLoss);                                    // :377
```

- **The scratch tensor takes the source's shape.** `setTensorValuesForConversion` sets `outputTensor->shape = originalTensor->shape` (`src/tensor/Tensor.c:291`). Every conversion walks `calcNumberOfElementsByTensor(inputTensor)` elements, for example `src/tensor/TensorConversion.c:438` (`convertFloatTensorToSymInt32Tensor`), `:510` (`convertSymInt32TensorToFloat32Tensor`), and the same-type `memmove` at `:2467-2470`.
  - **Longer loss or propLoss:** `:360` and `:367` write `count(loss)` or `count(propLoss)` floats into a buffer sized for `count(input)`. That is a stack-buffer-overflow write. For propLoss, `:377` also reads past the end.
  - **Shorter loss:** `softmaxVjpRows` walks the input's `rows * rowLen` elements, so it reads the uninitialized tail of `lossFloatData`. The result is wrong dx.
  - **Shorter propLoss:** the VJP stays in bounds, but `:377` converts back only `count(propLoss)` elements, so dx is silently truncated.
- **Zero-element path: the three VLA bounds are the only UB.** Everything after them already handles n = 0:
  - `softmaxVjpRows` returns on `rowLen == 0` (`:316-318`, from #456) and loops over zero rows for `[0, N]`.
  - `findAbsMaxFloat` returns 0.f for n = 0 (`src/arithmetic/MinMax.c:23-25`).
  - The conversions loop n times.
- **FLOAT32 arm.** `softmaxBackwardFloat` (`:337-343`) casts `loss->data` and `propLoss->data` to `float*` and walks the input's `rows * rowLen` elements:
  - Shorter loss: out-of-bounds read.
  - Shorter propLoss: out-of-bounds write.
  - Longer propLoss: the tail is left stale.
  - Longer loss: the tail is silently ignored.
- **The BFP arm's gate cannot be reused here.** The BFP arm does check the loss count (`:407`, `bfpRequireElementCount`). That helper takes a `bfpQConfig_t` and calls `validateBfpQConfigShape` (`src/arithmetic/include/BfpKernelSupport.h:99-109`), so it does not apply to FLOAT32 or SYM wires. The existing pattern for plain wires is the inline count check in `src/layer/Dropout.c:370-376`.

## Impact / reachability

- **Not reachable from the training loop.** `src/` calls a layer backward in one place only, `src/userApi/training_loop/calculate_grads/CalculateGradsSequential.c:118-127`. Both grads come from `initGradTensor`, which copies the layer-output shape (`:334-337`). `softmaxCalcOutputShape` (`Softmax.c:477-483`) copies the input shape, so the three counts always match. With CrossEntropy, the loop skips the last layer (`:94-97`), which is Softmax in CE models.
- **Reachable from direct callers** of `softmaxBackward` or `layerFunctions[SOFTMAX].backward` (unit tests now, ir2c-generated code later):
  - **SYM arm:** a longer loss or propLoss overflows the stack scratch. A shorter loss or propLoss gives silently wrong or truncated dx.
  - **FLOAT32 arm:** a shorter loss is read out of bounds and a shorter propLoss is written out of bounds, both in the caller's buffers. A longer propLoss leaves a stale dx tail.
  - None of these cases prints anything.
  - A zero-element input on the SYM arm is UB.

## Proposed fix

Add a static gate that both non-BFP arms call. It returns the shared element count:

```c
static size_t softmaxBackwardRequireWires(const tensor_t *input, const tensor_t *loss,
                                          const tensor_t *propLoss, size_t *rows, size_t *rowLen);
```

It does three things, in this order:

1. Calls `softmaxRowGeometry(input, ...)`, moved up from `:373`. The identity-order check then runs before any conversion.
2. Checks that `calcNumberOfElementsByTensor(loss)` and `calcNumberOfElementsByTensor(propLoss)` each equal the input's count. On a mismatch it calls `PRINT_ERROR` and `exit(1)`, as `Dropout.c:370-376` does.
3. Returns the input's count.

How the arms use it:

- **`softmaxBackwardFloat`** calls the gate in place of its own `softmaxRowGeometry` call.
- **`softmaxBackwardSymInt32`** calls the gate first, then `if (inputSize == 0) return;`. It declares its three VLAs only after that line.
  - The single `count == 0` check covers both `[2, 0]` and `[0, N]`, because `rowLen == 0` implies count 0.
  - The gate runs before the return, so an empty input with a non-empty loss still fails fast. That is stricter than the BFP kernel, which returns on `n == 0` (`:401-403`) before it reaches its gate.

Keep the gate out of the BFP arm. That arm accepts `propLoss == NULL` (`:448-450`) and has its own gate inside the kernel.

**Side effect for n = 0.** Today, an empty SYM_INT32 propLoss still passes through the conversion, which writes `scale = 1.f` (`TensorConversion.c:449-456`). The early return skips that write. With no elements the scale carries no information, so set it explicitly only if a test should pin it.

Optional, in the same change: remove `convertTensor(propLoss, &propLossFloat)` at `:367` and keep the `setTensorValuesForConversion` call. The conversion is wasted work, because `softmaxVjpRows` overwrites every element before `:377` converts the buffer back.

## Tests

- **`testSoftmaxBackwardSymEmptyRowsAreNoOp`**
  - The SYM_INT32 twin of `testSoftmaxBackwardEmptyRowsAreNoOp` (`test/unit/layer/UnitTestSoftmax.c:796`). Build the wires the way `unitTestSoftmaxBackwardSymInt32` does (`:213-242`), without a fill, for both `[2, 0]` and `[0, 3]`.
  - As with the FLOAT32 twin, only a sanitizer turns this test RED. It must run under `unit_test_asan` (`-fsanitize=address,undefined`, `CMakePresets.json:59`; CI builds it with clang at `ci.yml:188-194`).
  - `unit_test_ubsan` enables only `signed-integer-overflow,float-cast-overflow` (`:70`), so it does not catch `vla-bound`.
  - Mutations:
    - Removing `if (inputSize == 0) return;` in `softmaxBackwardSymInt32` turns the test RED through `vla-bound`.
    - Replacing that line with a `rowLen == 0` check turns the `[0, 3]` case RED.
- **Death tests**
  - Use `ASSERT_EXITS_WITH_FAILURE` (`test/unit/support/include/DeathTest.h:52`). Model them on `testSoftmaxBackwardBfpRejectsCountMismatchLoss` (`:1183`), with a SYM `[2, 3]` input.
  - **propLoss `[2, 2]` (shorter).** Without the gate, the run stays in bounds and reads only initialized values, so the child exits 0. RED on every preset. This is the deterministic discriminator.
  - **loss `[2, 2]` (shorter).** Without the gate, the VJP reads the uninitialized tail of `lossFloatData`. The outcome depends on the preset:
    - `unit_test` and `unit_test_debug`: the child exits 0. RED.
    - `unit_test_asan`: if the garbage is NaN or Inf, the child may instead abort on a UBSan float-cast-overflow in `roundHalfAway` (`src/arithmetic/Rounding.c:11-13`). Also RED.
    - `unit_test_ubsan` in CI (Linux, gcc; `UBSAN_OPTIONS` sets `halt_on_error=1` but not `abort_on_error`, `CMakePresets.json:263`): the same trap exits with status 1 and would satisfy the assertion. Run the mutation check on a non-sanitizer preset.
  - **loss `[2, 4]` and propLoss `[2, 4]` (longer).** Without the gate, the stack scratch overflows. Under `unit_test_asan` the child aborts (`abort_on_error=1`, `CMakePresets.json:245`), and the harness reports "terminated by signal". On plain presets the child either exits 0 or crashes. RED either way.
  - **FLOAT32 arm:** loss `[2, 2]` (out-of-bounds read) and propLoss `[2, 2]` (out-of-bounds write), with heap wires built by `initTensor`. This path has no float-to-int cast, so both tests are RED on every preset.
  - Mutation: delete the gate and run on `unit_test` or `unit_test_debug`. Every death test turns RED.

## Found during

- (a): #152 design spec §8, pre-existing and deferred by ruling R4.
- (b): Codex review of #152 PR 2.

Refs #152
